app/src/androidTest/java/org/mifos/mobile/ui/activities/LoginSetPasscodeTest.kt

```diff
@@ -0,0 +1,192 @@
+package org.mifos.mobile.ui.activities
+
+import android.view.View
+import android.view.ViewGroup
+
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.IdlingRegistry
+import androidx.test.espresso.ViewInteraction
+import androidx.test.espresso.action.ViewActions.*
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.matcher.ViewMatchers.*
+import androidx.test.filters.LargeTest
+import androidx.test.rule.ActivityTestRule
+import androidx.test.rule.GrantPermissionRule
+import androidx.test.runner.AndroidJUnit4
+
+import org.hamcrest.Description
+import org.hamcrest.Matcher
+import org.hamcrest.Matchers.`is`
+import org.hamcrest.Matchers.allOf
+import org.hamcrest.TypeSafeMatcher
+
+import org.junit.After
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mifos.mobile.R
+import org.mifos.mobile.utils.EspressoIdlingResouce
+
+/**
+ * UI Testing for Login using Espresso
+ *
+ * @author Ashwin Ramakrishnan
+ * @since 22/08/2020
+ */
+
+@LargeTest
+@RunWith(AndroidJUnit4::class)
+class LoginSetPasscodeTest {
+
+    @Rule
+    @JvmField
+    var mActivityTestRule = ActivityTestRule(SplashActivity::class.java)
+
+    @Rule
+    @JvmField
+    val grantPermissionRule: GrantPermissionRule =
+        GrantPermissionRule.grant(android.Manifest.permission.READ_PHONE_STATE)
+
+    @Before
+    fun registerIdlingResource() {
+        IdlingRegistry.getInstance().register(EspressoIdlingResouce.countingIdlingResource)
+    }
+
+    @After
+    fun unregisterIdlingResource() {
+        IdlingRegistry.getInstance().unregister(EspressoIdlingResouce.countingIdlingResource)
+    }
+
+    @Test
+    fun loginSetPasscodeTest() {
```

Comment:
Is it an integration test for a single flow, clarify. If not split it multiple flows into multiple tests, instead of covering all flows in a single test method.

Reply:
This test is meant to cover the login flow completely. So I have included steps to set passcode too so that the complete flow is covered.

Reply:
Sounds good.
Let's go with a failure case, where login is a failure. `Invalid creds`

Suggested naming convention, add suffix `_validCreds` and `_invalidCreds`. You can keep the prefix same for both the test methods.

Reply:
Yes, will make those changes